app/uk/gov/hmrc/fileupload/s3/S3JavaSdkService.scala

```diff
@@ -248,26 +248,31 @@ class S3JavaSdkService(configuration: com.typesafe.config.Config, metrics: Metri
   ): Future[ZipData] = {
     val fileName = s"$envelopeId.zip"
     val tempFile = TemporaryFile(prefix = "zip")
+    val (uploadFinished, md5Finished) =
+      broadcast2(
+        source = zipSource(envelopeId, files),
+        sink1  = FileIO.toPath(tempFile.file.toPath),
+        sink2  = Md5Hash.md5HashSink
+      ).run()
     (for {
-       _            <- zipToFile(envelopeId, files, tempFile.file)
+       _            <- Future.successful(Logger.debug(s"zipping $envelopeId to ${tempFile.file}"))
+       _            <- uploadFinished
+       md5Hash      <- md5Finished
        fileSize     =  tempFile.file.length
-       is           =  new java.io.FileInputStream(tempFile.file)
-
-       // decorate inputstream so we can calculate checksum on same pass
-       md           =  MessageDigest.getInstance("MD5")
-       dis          =  new DigestInputStream(is, md)
-
+       _            =  Logger.debug(s"uploading $envelopeId to S3")
        uploadResult <- uploadFile(
                          bucketName = awsConfig.transientBucketName,
                          key        = S3Key.forZipSubdir(awsConfig.zipSubdir)(fileName),
-                         file       = dis,
+                         file       = new java.io.FileInputStream(tempFile.file),
                          metadata   = { val om = new ObjectMetadata()
                                         om.setSSEAlgorithm(SSEAlgorithm.KMS.getAlgorithm)
                                         om.setContentLength(fileSize)
                                         om.setContentType("application/zip")
+                                        om.setContentMD5(md5Hash)
```

Comment:
this is the motivation: generating the md5Hash earlier, and getting aws confirmation of the value.